frontend/src/Components/ScaleSelector/ScaleSelector.tsx

```diff
@@ -1,25 +1,128 @@
+import { useEffect, useRef, useState } from 'preact/hooks';
 import { ScaleName, SCALES } from '../../../../shared/scales';
 import { SELECT_CHANGE_SCALE } from '../../constants';
+import sharedClasses from '../../styles.module.css';
 import { connectToWebSocket } from '../WebSocket/WebSocket';
 import classes from './ScaleSelector.module.css';
 
-export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => (
-  <select
-    disabled={!connected}
-    name="scale"
-    class={classes.select}
-    onChange={({ target }) =>
-      setScale(SCALES[(target as HTMLSelectElement).value as ScaleName].values)
+const availableScales = Object.entries(SCALES);
+
+export const ScaleSelector = connectToWebSocket(({ socket: { connected, setScale } }) => {
+  const [open, setOpen] = useState(false);
+  const [selected, setSelected] = useState(-1);
+  const selectionButtonRef = useRef<HTMLButtonElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+
+  const close = () => {
+    setOpen(false);
+    setSelected(-1);
+  };
+
+  const toggle = () => {
+    if (open) {
+      close();
+    } else {
+      setSelected(0);
+      setOpen(true);
+    }
+  };
+
+  const handleBlur = ({ relatedTarget }: FocusEvent) => {
+    if (relatedTarget !== selectionButtonRef.current && relatedTarget !== dropdownRef.current) {
+      close();
+    }
+  };
+
+  const handleKeyboardNavigation = (event: KeyboardEvent) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        setSelected((selected + 1) % availableScales.length);
+        event.preventDefault();
+        break;
+      case 'ArrowUp':
+        setSelected((Math.max(selected, 0) + availableScales.length - 1) % availableScales.length);
+        event.preventDefault();
+        break;
+      case 'Home':
+      case 'PageUp':
+        setSelected(0);
+        break;
+      case 'End':
+      case 'PageDown':
+        setSelected(availableScales.length - 1);
+        break;
+      case 'Escape':
+        close();
+        selectionButtonRef.current?.focus();
+        break;
+      case 'Enter':
+      case ' ':
```

Comment:
Did you consider using `event.code` for the switch instead of `event.key`? This would allow to use `Space` here, which would improve readability for me.

Reply:
`code` makes me slightly uneasy as it "does not take keyboard layout into account". E.g. on my keyboard, "ö" produces the code "Semicolon" and "-" produces "Slash". But the for the keys we are interested in it does not appear to make a difference so yes, we can do that too.

Reply:
Done